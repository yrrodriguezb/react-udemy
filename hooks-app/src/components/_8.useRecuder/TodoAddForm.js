import React from 'react'
import { useForm } from '../../Hooks/useForm';
import PropTypes from "prop-types";

export const TodoAddForm = ({ handleAddTodo }) => {

  const [ formValues, handleInputChange, reset ] = useForm({
    description: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (formValues.description.trim().length == 0)
      return;
    
    const newTodo = {
      id: new Date().getTime(),
      desc: formValues.description ,
      done: false
    }

    handleAddTodo(newTodo);
    reset();
  }

  return (
    <React.Fragment>
      <h4 className='fs-3'>Agregar Todo</h4>
      <hr />

      <form onSubmit={ handleSubmit }>
        <div className="mb-3">
          <label 
            htmlFor="description" 
            className="form-label"
          >
            Descripción
          </label>

          <input 
            type="text" 
            name='description' 
            className="form-control" 
            id="description" 
            aria-describedby="descHelp"  
            value={ formValues.description }
            onChange={ handleInputChange }
          />

          <div 
            id="descHelp" 
            className="form-text"
          >
            Descripción de la tarea.
          </div>
        </div>
        
        <button 
          type="submit" 
          className="btn btn-primary"
        >
          Guardar
        </button>
      </form>
    </React.Fragment>
  )
}

TodoAddForm.propTypes = {
  handleAddTodo: PropTypes.func
}
