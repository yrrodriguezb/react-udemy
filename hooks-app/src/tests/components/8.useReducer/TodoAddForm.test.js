import React from 'react';
import { shallow } from 'enzyme';
import { TodoAddForm } from '../../../components/_8.useRecuder/TodoAddForm';
import '@testing-library/jest-dom';

describe('Pruebas en <TodoAddForm />', () => {
  const handleAddTodo = jest.fn();

  const wrapper = shallow(
    <TodoAddForm 
      handleAddTodo={ handleAddTodo }
    />
  )

  test('Debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('No debe llamar el handleAddTodo', () => {
    // Otra forma de obtener una propiedad, en este caso la funcion que controla el submit del form
    const handleSubmit = wrapper.find('form').prop('onSubmit');

    handleSubmit({
      preventDefault: () => {}
    });

    // Valida que la funcion no se llame en ninguna vez
    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('Debe llamar el handleAddTodo', () => {
    const value = 'Aprender React';
    
    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description'
      }
    });

    // Otra forma de obtener una propiedad, en este caso la funcion que controla el submit del form
    const handleSubmit = wrapper.find('form').prop('onSubmit');

    handleSubmit({
      preventDefault: () => {}
    });

    // Valida que la funcion se llame una vez
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    
    // Valida que la funcion se llame una vez con un objeto como parametro
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    
    // Valida que la funcion se llame con el objeto adecuado y sus valores esperados
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false
    });

    // Valida que despues de agregar un todo la descripcion del formulario quede vac√≥o, (Funcion reset)
    // Se valida el efecto de llamar la funcion "reset", no se valida que se llame dicha funcion
    expect(wrapper.find('input').prop('value')).toBe('');
  });
});