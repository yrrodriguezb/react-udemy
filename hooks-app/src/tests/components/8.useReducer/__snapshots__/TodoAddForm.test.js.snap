// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAddForm /> Debe mostrarse correctamente 1`] = `
<Fragment>
  <h4
    className="fs-3"
  >
    Agregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="description"
      >
        Descripción
      </label>
      <input
        aria-describedby="descHelp"
        className="form-control"
        id="description"
        name="description"
        onChange={[Function]}
        type="text"
        value=""
      />
      <div
        className="form-text"
        id="descHelp"
      >
        Descripción de la tarea.
      </div>
    </div>
    <button
      className="btn btn-primary"
      type="submit"
    >
      Guardar
    </button>
  </form>
</Fragment>
`;
